### PR TITLE
[INFR-1205] add support for branch/master envars with tests

### PR DESCRIPTION
**Pull request by aks**
This PR adds two new options: `-e ENVAR=VALUE` and `-E ENVAR=VALUE`, which can be provided multiple times to declare envars and values to be assigned for the test branch and the reference branch, respectively.

It also adds a `-M` option as a short-hand for `--migrations`.

This allows perf_check to test changes that depend on specific envar values.

There is a new example test to verify the use of the envar support at lines 52-74 in `server_spec.rb`.

All specs pass.

### CODE REVIEW

**Comment by aks** [comment]
The TravisCI failure is in the master branch -- which suggests that the test_app is missing the "railties" gem.   Not sure how to fix the "master" branch?

File: lib/perf_check.rb [hunk 1]
```diff
@@ -107,14 +108,14 @@ def run
...
     reference_test.switch_to_reference_context
-    profile_test_case(reference_test)
+    profile_test_case(reference_test, reference: true)
   end
 
-  def profile_test_case(test)
+  def profile_test_case(test, index: nil, reference: false)
     trigger_before_start_callbacks(test)
-    server.restart
+    server.restart(reference: reference) unless index == 0 || options.diff
```
Comment (siggymcfried) [inline lib/perf_check.rb]: Are we intentionally regressing https://github.com/rubytune/perf_check/pull/19?

Comment (aks) [reply lib/perf_check.rb]: No, the conditional was based on the logic that predated #19.   I'll remove the `diff` conditional.  However, the `index == 0` logic seems reasonable given that the server is started up automatically before the first iteration, so restarting the server before the first iteration seems unnecessary.

Comment (siggymcfried) [reply lib/perf_check.rb]: That unnecessary restart was also handled in #19

File: lib/perf_check/server.rb [hunk 2]
```diff
@@ -66,7 +66,7 @@ def latest_profiler_url
     end
 
     def profile
-      http = Net::HTTP.new(host, port).tap{ |http| http.read_timeout = 1000 }
+      http = Net::HTTP.new(host, port).tap{ |ht| ht.read_timeout = 1000 }
```
Comment (sudara) [inline lib/perf_check/server.rb]: Intentional?

File: lib/perf_check.rb [hunk 3]
```diff
@@ -69,18 +69,19 @@ def add_test_case(route)
   def run
     begin
       run_migrations_up if options.run_migrations?
+      server.restart(reference: false)
```
Comment (siggymcfried) [inline lib/perf_check.rb]: Intentional?

File: spec/perf_check_spec.rb [hunk 4]
```diff
@@ -70,7 +70,6 @@
     end
 
     it "should ensure to run migrations down if options.run_migrations?" do
-      perf_check.add_test_case('/xyz')
```
Comment (siggymcfried) [inline spec/perf_check_spec.rb]: Intentional?

**Comment by sudara** [comment]
A test case knows when its context is `:reference` https://github.com/rubytune/perf_check/blob/master/lib/perf_check/test_case.rb#L119

Maybe we can ask it here and send the appropriate env var into `server.restart` https://github.com/rubytune/perf_check/blob/master/lib/perf_check.rb#L115

File: lib/perf_check/server.rb (hunk 2 above)
Comment (aks) [reply lib/perf_check/server.rb]: When running tests, this showed up as a warning -- the masking of an outer variable by an inner variable of the same name.    So, I changed the inner variable to remove the warning.

File: lib/perf_check.rb (hunk 3 above)
Comment (aks) [reply lib/perf_check.rb]: yes.  specs don't pass unless the server is restarted after the migrations.

File: spec/perf_check_spec.rb (hunk 4 above)
Comment (aks) [reply spec/perf_check_spec.rb]: Yes. With this line, the spec fails.  Without it, the spec passes.

In addition, this spec is about ensuring that migrations are run when the option is set; the next test (below) has the test case comparison.

File: lib/perf_check.rb (hunk 3 above)
Comment (siggymcfried) [reply lib/perf_check.rb]: It seems to be undoing https://github.com/rubytune/perf_check/pull/19/files

File: lib/perf_check/server.rb (hunk 2 above)
Comment (sudara) [reply lib/perf_check/server.rb]: Ah, gotcha!

File: lib/perf_check.rb (hunk 3 above)
Comment (aks) [reply lib/perf_check.rb]: So, I pulled `rubytune/master` and merged, and found some recent changes there in conflict.  I'll push again after resolving the conflicts, and accommodating Sudara's suggestions, too.

File: spec/perf_check_spec.rb (hunk 4 above)
Comment (aks) [reply spec/perf_check_spec.rb]: This change was obviated from the recent merge from `rubytune/master`.  Updates pending.

File: lib/perf_check.rb (hunk 3 above)
Comment (aks) [reply lib/perf_check.rb]: This change is obviated from a recent merge with `rubytune/master`.  Updates pending..

File: lib/perf_check.rb [hunk 5]
```diff
@@ -104,15 +104,15 @@ def run
...
     reference_test = test_cases[1]
-    profile_test_case(first)
+    profile_test_case(first, reference: false)
     reference_test.switch_to_reference_context
-    profile_test_case(reference_test)
+    profile_test_case(reference_test, reference: true)
   end
 
-  def profile_test_case(test)
+  def profile_test_case(test, index: nil, reference: false)
```
Comment (siggymcfried) [inline lib/perf_check.rb]: Index is unused

Comment (aks) [reply lib/perf_check.rb]: Thanks.  That's a left-over artifact of the code before your most recent PR.  I'll remove it.

**Comment by aks** [comment]
@siggymcfried:
* in regard to the new `Rakefile` -- yes, without it, it's not possible to do `bundle exec rake spec`.
* in regard to the `rails_helper.rb` -- without it, we get errors in running `rake spec` without it _(even though it's empty)_

File: lib/perf_check/server.rb [hunk 6]
```diff
@@ -110,22 +116,31 @@ def start
...
 
+      # setup envars appropriate to the branch or the reference
+      if reference.nil?
+        reference = @last_reference || false
+      else
+        @last_reference = reference
+      end
+      envs = reference ? perf_check.options.reference_envs : perf_check.options.branch_envs
+      @last_reference = reference
+      (envs || {}).each_pair { |var, val| ENV[var] = val }
```
Comment (siggymcfried) [inline lib/perf_check/server.rb]: Is envs ever nil? What is the check we have at the beginning doing?

File: bin/perf_check [hunk 7]
```diff
@@ -18,7 +18,9 @@ app_root =
 
 
 if File.exists?("#{app_root}/tmp/pids/server.pid")
-  abort("It looks like a rails server is already running. Shut it down or remove the pid in tmp/pids/server.pid before continuing with perf_check.")
+  $stderr.puts "It looks like a rails server is already running. Shut it down or remove the pid in tmp/pids/server.pid before continuing with perf_check."
+  $stderr.puts `ps axuww`.split(/\n/).grep(/perf_check/).join('\n')
```
Comment (siggymcfried) [inline bin/perf_check]: What does this do? Can we move it to a method with a descriptive name?

File: lib/perf_check.rb [hunk 8]
```diff
@@ -128,6 +129,30 @@ def profile_test_case(test, reference: false)
...
+
+    # setup envars appropriate to the branch or the reference
+    if reference.nil?
+      reference = @last_reference || false
+    else
+      @last_reference = reference
+    end
+    test_envs = reference ? options.reference_envs : options.branch_envs
+    @last_reference = reference
+    (test_envs || {}).each_pair { |var, val| env[var] = val }
```
Comment (siggymcfried) [inline lib/perf_check.rb]: When is test_envs nil or false?

File: lib/perf_check/server.rb [hunk 9]
```diff
@@ -94,53 +104,69 @@ def profile
...
+    end
 
-      # setup envars appropriate to the branch or the reference
-      if reference.nil?
-        reference = @last_reference || false
-      else
-        @last_reference = reference
+    # kill a process, first with QUIT, then with KILL
+
+    def kill_process(pid)
```
Comment (siggymcfried) [inline lib/perf_check/server.rb]: Is this all related?

Comment (aks) [reply lib/perf_check/server.rb]: Yes. Testing uncovered a flaky test that sometimes fails because the previous server had not yet shutdown, resulting in the subsequent process failing to start because there was already a previous server running.
